sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,727 @@
+
+                # Save recently generated leaves
+                leaves_dict[child_id] = {
+                    "samples": samples_mask,
+                    "error_or_size": metrics_values[picked_id],
+                }
+
+            # Split cluster is no longer leaf
+            del leaves_dict[biggest_id]
+
+            last_center_id += 2
+
+            # Pick new 'biggest cluster' to split
+            biggest_id, _ = max(
+                leaves_dict.items(), key=lambda x: x[1]["error_or_size"]
+            )
+
+        # Delete Initial cluster
+        del tree_dict[-1]
+
+        # Sort Leaf centers by their root
+        centers_hierarchical = [0, 1]
+        while len(centers_hierarchical) != self.n_clusters:
+            nested_centers = []
+
+            for center_id in centers_hierarchical:
+                center_ids = (
+                    [center_id]
+                    if tree_dict[center_id]["children"] is None
+                    else tree_dict[center_id]["children"]
+                )
+                nested_centers.extend(center_ids)
+
+            centers_hierarchical = nested_centers
+
+        # Initialize Labels
+        labels = np.full(X.shape[0], -1, dtype=np.intc)
+
+        centers = []
+
+        for i, center_id in enumerate(centers_hierarchical):
+            # Save cluster centers in hierarchical order
+            centers.append(tree_dict[center_id]["center"])
+
+            # Assign labels to proper data points
+            labels[leaves_dict[center_id]["samples"]] = i
+
+        centers = np.asarray(centers)
+
+        # Inner Tree will be later used at 'predict' method:
+        self._inner_tree = tree_dict
+
+        return centers, labels
+
+    def predict(self, X, sample_weight=None):
+        """Predict which cluster each sample in X belongs to by going down
+        the hierarchical tree in searching of closest leaf cluster.
+
+        In the vector quantization literature, `cluster_centers_` is called
+        the code book and each value returned by `predict` is the index of
+        the closest code in the code book.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            New data to predict.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X. If None, all observations
+            are assigned equal weight.
+
+        Returns
+        -------
+        labels : ndarray of shape (n_samples,)
+            Index of the cluster each sample belongs to.
+        """
+        check_is_fitted(self)
+
+        X = self._check_test_data(X)
+        x_squared_norms = row_norms(X, squared=True)
+        sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+
+        # Note: Well optimized Bisect K-Means should produce clusters with same or
+        # even lower inertia than K-Means.
+        # So after optimization part below may be replaced
```

Comment:
I don't understand this comment, which part can be replaced?

Reply:
While writing I was thinking on part of code below this comment, which was responsible for assigment to nearest cluster from tree, with assumption that optimized `BisectKMeans` in theory might produce centroids with lower inertia. In that case simple `_check_labels` function should also do the work properly.
I remove that comment, since I'm not sure if moving that part into cython wouldn't provide even better performance.